### ISSUE TITLE
Allow null values for resultvalues tables

### ISSUE THREAD

**Issue by miguelcleon** [ODM2/ODM2#110]
Null values should be allowed for many columns in result values tables. These tables are ODM2Results.ProfileResults, ODM2REsults.PointCoverageResults, ODM2Results.SectionResults, ODM2Results.TransectResults, ODM2REsults.TrajectoryResults, ODM2Results.SpectraResults. 

For Instance I made the following changes on my profileresultvalues table. Someone might not know the exact date and time that some soil sample was taken, and a time aggregation interval may not be applicable. Maybe the Zlocation, and zlocationunitsid should still be NOT NULL. 

ALTER TABLE odm2.profileresultvalues
   ALTER COLUMN datavalue DROP NOT NULL;
ALTER TABLE odm2.profileresultvalues
   ALTER COLUMN valuedatetime DROP NOT NULL;
ALTER TABLE odm2.profileresultvalues
   ALTER COLUMN valuedatetimeutcoffset DROP NOT NULL;
ALTER TABLE odm2.profileresultvalues
   ALTER COLUMN zlocation DROP NOT NULL;
ALTER TABLE odm2.profileresultvalues
   ALTER COLUMN zaggregationinterval DROP NOT NULL;
ALTER TABLE odm2.profileresultvalues
   ALTER COLUMN zlocationunitsid DROP NOT NULL;
ALTER TABLE odm2.profileresultvalues
   ALTER COLUMN timeaggregationinterval DROP NOT NULL;
ALTER TABLE odm2.profileresultvalues
   ALTER COLUMN timeaggregationintervalunitsid DROP NOT NULL;

**Comment by miguelcleon** [comment]
Probably not a good idea for me to change this on my end. In any case I think these tables need to allow more null values.

**Comment by emiliom** [comment]
@miguelcleon, I'll start with these references: The ODM2 manuscript that we've submitted for publication (currently in revision) touches on some of these issues. Also, there are relevant ODM2 wiki entries:
- [Best Practices: Handling Dates and Times](https://github.com/ODM2/ODM2/blob/master/doc/ODM2Docs/bp_handlingdatesandtimes.md). It focuses in detail mainly on sensor time series; we should add specimen-based examples.
- [Best Practices: Encoding Data for which the Exact Date is Unknown](https://github.com/ODM2/ODM2/blob/master/doc/ODM2Docs/bp_unknowndates.md)
- [Results - Depth Profile Coverage Result Type](https://github.com/ODM2/ODM2/blob/master/doc/ODM2Docs/ext_results_profile.md)

I think the broader issue you're facing involves a distinction between having exact information and being able to provide a reasonable estimate and an annotation about the precision of that estimate.

To use your example of a soil sample in a profile, I can totally see that the actual collection day, much less the time, may not have been recorded (specially for older sample collections). But if, for example, the month is known, a `valuedatetime` should still be entered; ideally ODM2 should have an optional mechanism to specify the precision of that estimate. The [recommended best practice](https://github.com/ODM2/ODM2/blob/master/doc/ODM2Docs/bp_unknowndates.md) involves creating a `ProfileResultValueAnnotations` entry, though that can be awkward.

Another thing to keep in mind is that there's a chain of related timestamps throughout ODM2. Let's take the example of a soil sample (part of a larger profile collection) collected on 2000-2-17, analyzed for organic C & N content 3 months later, and for mineral elemental composition 1 year later:
- Sample collection date should be captured in the `Actions` table. This action does not produce a `Results` entry, but has its own sample collection method description.
- If there are additional sample processing steps, and you want to capture them in your ODM2 database, they can be separate, related (children) Actions with their own methods.
- Each specific analysis method would have its own action, with associated `Results` entries each with its own `ResultDateTime` (which currently can be null) and `ProfileResultValues.ValueDateTime`. Frankly, I can't remember if the our intent (or best practice) was to have this `ValueDateTime` (and all similar columns for other specimen result types) correspond to the analysis timestamp or the sample collection timestamp (where the latter is the one that's more scientifically interesting). I think it's the former, but hopefully @horsburgh can clarify; the [ODM2 DbWrench schema diagram doesn't have any comments for ValueDateTime fields](http://odm2.github.io/ODM2/schemas/ODM2_Current/diagrams/ODM2Results.html).

In this scenario, it's likely that the analysis timestamps are known, though possibly only to the day without the actual time; and the soil collection sample `Actions` timestamp is known only to the month, but that timestamp is `Actions.BeginDateTime`, not `ProfileResultValues.ValueDateTime`. As [described here](https://github.com/ODM2/ODM2/blob/master/doc/ODM2Docs/bp_unknowndates.md), the recommended best practice would be to indicate this by entering a timestamp with the first day of the month: "2000-2-1 12:00 AM"

In general, I think it's very risky to make such important ResultValues fields nullable, though some may not be as important as others. We should try to accommodate the valid needs instead via appropriate CV terms (eg, for `ProfileResults.AggregationStatisticCV`), best practices for imprecise timestamps and annotations, clarifications of best practices for the timestamp "chain", etc; likewise for other dimensions or axes (Z, etc). 

@horsburgh, what do you think? It would also be helpful to hear from the EarthChem/IEDA group, since this situation must be even more common with their geological samples.

**Comment by miguelcleon** [comment]
Ok, I can see keeping the valuedatetime and zaggregationinterval as non-nullable but how about the timeaggregationinterval and timeaggregationintervalunitsid? Those don't seem to apply in my use case. I suppose I could enter a place holder value of 0 and units of undefined or none.

**Comment by emiliom** [comment]
I agree that timeaggregationinterval and timeaggregationintervalunitsid are not terribly meaningful for soil sample results. Your solution seems sensible (0 and undefined/none). Alternatively you could enter other values that are also place holders but might be a bit more physically meaningful (eg, 1 hour), though made up.

**Comment by miguelcleon** [comment]
I think I will go with 0 and and a units of 'not measured' for now.
